text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,673 @@
+        * Slot content - default and named slots
+    * Static resource
+        * Image urls
+        * Javascript resource
+        * CSS resource
+    * Dynamic components
+    * Template directives
+        * for:each
+        * if
+        * iterator
+        * key
+        * lwc:dom="manual"
+    * Event handlers
+    * Slots
+* Javascript file
+    * Component classes
+        * Inheritance
+        * Public & private properties
+        * Decorators
+        * getters/setters
+        * Public methods
+    * Imports
+        * Builtin modules
+        * External modules
+        * Salesforce scoped modules
+        * Other modules(components and libraries)
+        * Relative imports
+        * Dynamic imports
+    * Exports
+        * Default exports
+        * Named exports
+        * Aggregating modules
+    * JSDoc
+        * Class
+        * Properties
+        * Public methods
+    * Programmatically generated events
+        * type
+        * options: composed, bubbles
+    * Programmatically fetched static resources
+        * `<script>`
+        * `<img>`
+        * `<link>`
+* CSS file
+    * Tokens
+
+## Metadata shape
+
+```json
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "description": "A representation of metadata gathered for a LWC bundle",
+    "type": "object",
+    "properties": {
+        "name": { "type": "string" },
+        "namespace": { "type": "string" },
+        "moduleType": { 
+            "type": "string",
+            "enum": [ "Component", "Library", "CssOnly" ]
+        },
+        "templates": { 
+            "type": "array",
+            "items": { "$ref": "#/definitions/HTMLTemplate"}
+        },
+        "scripts": {
+            "type": "array",
+            "items": { "$ref": "#/definitions/Script"}
+        },
+        "defaultComponentClass": {
+            "type": "array",
+            "items": { 
+                "$ref": "#/definitions/Class"
+            }
+        },
+        "css": {
+            "$ref": "#/definitions/CSSFile"
+        }
+    },
+    "definitions": {
+        "HTMLTemplate": {
+            "type": "object",
+            "properties": {
+                "fileType": { "$ref": "#/definitions/FileType" },
+                "fileName": { "type": "string" },
+                "children": {
+                    "type": "array",
+                    "items": { 
+                        "$ref": "#/definitions/CustomElementReference"
+                    }
+                },
+                "staticResources": {
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/definitions/StaticResourceReference"
+                    }
+                },
+                "dynamicChildren": {
+                    "type": "array",
+                    "items": { 
+                        "$ref": "#/definitions/DynamicComponentReference"
+                    }
+                },
+                "directives": {
```

Comment:
After thinking it more, and having debated this with @pmdartus in the past, I think we should just end up exposing a "pseudo AST": Doesn't contain all the nodes, but the nodes that has a meaning. ex: a node with an expression, directive, or any custom element.

Reply:
Also anything that contains a url will be interesting

Reply:
Agree. The LNS team is already consuming our bindingAST, exposed as part of template metadata. They are also open for consuming an AST for js files as well.